design/one-pager-terrajet-metadata-extraction.md

```diff
@@ -0,0 +1,272 @@
+with a common structure. Our assumption is that Terraform registry website is
+also generated using these markdown files:
+- For `terraform-provider-aws`:
+  https://github.com/hashicorp/terraform-provider-aws/tree/main/website/docs/r
+- For `terraform-provider-azurerm`:
+  https://github.com/hashicorp/terraform-provider-azurerm/tree/main/website/docs/r
+- For `terraform-provider-google`:
+  https://github.com/hashicorp/terraform-provider-google/tree/main/website/docs/r
+
+Thus a common metadata scraper implementation can extract metadata from these
+well-formatted per-resource markdown documents. Any spotted errors can then
+potentially be corrected manually in the scraped YAML metadata document.
+Scrapers can optionally be chained: If desired, another scraper can append
+example HCL configurations read from a different source (such as the `examples`
+folder found in some of the Terraform native provider repositories as discussed
+above).
+
+For most Terraform native providers, we anticipate that Terraform registry
+scrapers will **not** run on HTTP, as the resource markdown files are part of
+their corresponding provider repositories. They can just read those markdowns
+from a pointed directory in the local filesystem, which is specified as a
+command-line argument, for instance. 
+
+As already indicated, if it turns out that a common registry scraper
+implementation is not suitable for a specific Terraform native provider, then a
+new scraper can be written as long as it produces metadata output in the
+expected metadata format by the Terrajet pipelines. Or even, if the cost of
+writing a new scraper is higher than manually authoring the metadata YAML
+documents (e.g., the number of resources in the native provider is small), we
+can just prepare the metadata YAML(s) by hand, just like the Terraform community
+manually maintains the corresponding markdown documents for the Terraform
+registry.
+
+### Terrajet Codegen Pipelines Consuming Metadata
+As [implemented][terrajet-pr-173] in the context of the [Terrajet issue #48] for
+example manifest generation for the big three providers, we could have some
+configurable codegen pipelines that consume the YAML resource metadata file(s)
+and produce example manifests, CRD documentation, etc. (as discussed above). The
+configured pipelines should not fail if the necessary metadata is missing: For
+instance, the [example manifest generation pipeline] should simply not generate
+an example manifest for a managed resource, if no sample HCL configuration is
+available for the corresponding Terraform resource in the metadata. Or, the
+metadata-enhanced CRD generation pipeline should simply skip doc comments if
+none or some are not available in the corresponding metadata document. 
+
+Metadata is valuable; the scrapers should capture as much metadata as possible
+and store them in the common format, even for future use cases we do not yet
```

Comment:
I'm a bit on the fence about storing everything we can because we have to have a format for each of those information that we don't use and once we do use them, we may want to change the format, i.e. where it lives in the YAML and also since we provide manual input ability, the format become one of our public APIs. So, I wonder if we should, for example, just keep the examples since that's all we use today and then once we use other information, we could add them.

Reply:
I have added example use cases for the proposed fields. In fact (apart from documentation) most have been utilized in provider-jet-azure. I think this also serves as an opportunity to show the potential we have here.